### ISSUE TITLE
DNS SOA record fields set failure 

### ISSUE THREAD

**Issue by vertexclique** [mfontanini/libtins#48]
DNS SOA records couldnt set also length couldnt calculated correctly

* primary name server
* responsible authority mailbox
* serial number
* refresh interval
* retry interval
* expire limit
* minimum ttl

You can see in below image

![soarecords](https://cloud.githubusercontent.com/assets/578559/5821496/4f4ad0b6-a0d5-11e4-8b11-8ed4e9d1276f.PNG)

**Comment by vertexclique** [comment]
@mfontanini

**Comment by mfontanini** [comment]
SOA records are not parsed, but you should still be able to perform a record.data() and get the binary representation of the data. Is this giving you wrong results?

**Comment by vertexclique** [comment]
I am trying to craft and i cannot craft because of that reasons… many libraries implemented this also inplemented in programming language level…

----
Mahmut Bulut

**Comment by mfontanini** [comment]
Yes, this should be implemented. I'll add this when I have some time.

Sorry for the inconvenience

**Comment by roee88** [comment]
I needed the same earlier today and wrote this code to encode strings like "ns.example.com. hostmaster.example.com. 2003080800 1d 15m 3w 2h". The encoded string can be passed to libtins in the resource record. Works well for my needs.

	int parse_time(std::string time_str)
	{
		static std::map<char, int> secs { {'s', 1}, { 'm', 60 }, { 'h',3600 }, { 'd',86400 }, { 'w',604800 }};

		char unit = ::tolower(time_str.back());
		if (secs.find(unit) != secs.end())
		{
			time_str.erase(time_str.end() - 1);
			return std::stoi(time_str) * secs[unit];
		}
		return std::stoi(time_str);
	}

	string encode_soa_data(const string& soa_string)
	{
		byte_array bytes;
		std::istringstream iss(soa_string);
		std::string value;

		// add mname and rname
		for (int i = 0; i < 2; ++i)
		{
			iss >> value;
			if (value.back() == '.')
				value.erase(value.end() - 1);

			std::string encoded_value = DNS::encode_domain_name(value);
			bytes.insert(bytes.end(), encoded_value.begin(), encoded_value.end());
		}
		int added = bytes.size();

		// prepare memory for time integers
		bytes.insert(bytes.end(), sizeof(uint32_t) * 5, 0);

		// add times
		for (int i = 0; i < 5; ++i)
		{
			iss >> value;
			uint32_t encoded_int = Endian::host_to_be(parse_time(value));
			std::memcpy(bytes.data() + added, &encoded_int, sizeof(uint32_t));
			added += sizeof(uint32_t);
		}

		return std::string(bytes.begin(), bytes.end());
	}

**Comment by mfontanini** [comment]
Okay, I added this to the develop branch. Note that I also changed some names, like DNS::Query is now DNS::query, DNS::Resource is DNS::resource and DNS::query::type is not DNS::query::query_type (so it's uniform with query_class). I added a typedef so Query still exists (and type() also), but they're both deprecated, so you will probably get some warnings from your compiler.

See the DNS::soa_record in dns.h to read the documentation on it.
